File: core/modules/purchase/purchase.service.ts
import { IPurchaseRepository, CreatePOData, UpdatePOData, PurchaseOrderQueryOptions } from './purchase.repository.interface.js';

let purchaseRepo: IPurchaseRepository;
let auditService: any;

export function initPurchaseService(repo: IPurchaseRepository, audit: any) {
    purchaseRepo = repo;
    auditService = audit;
}

const VALID_STATUSES = new Set(['pending', 'received', 'cancelled']);
const VALID_SORT_FIELDS = new Set(['id', 'supplier_name', 'order_date', 'status', 'item_count', 'total_cost']);

function httpError(message: string, statusCode: number): Error & { statusCode: number } {
    const error = new Error(message) as Error & { statusCode: number };
    error.statusCode = statusCode;
    return error;
}

function assertPositiveInteger(value: number, fieldName: string) {
    if (!Number.isInteger(value) || value <= 0) {
        throw httpError(`${fieldName} must be a positive integer`, 400);
    }
}

export function getAllPurchaseOrders(options: PurchaseOrderQueryOptions = {}) {
    const page = options.page ?? 1;
    const limit = options.limit ?? 10;
    const searchTerm = options.searchTerm ?? '';
    const status = options.status ?? '';
    const fromDate = options.fromDate ?? '';
    const toDate = options.toDate ?? '';
    const sortBy = options.sortBy ?? 'order_date';
    const sortOrder = (options.sortOrder ?? 'DESC').toUpperCase();

    assertPositiveInteger(page, 'page');
    if (!Number.isInteger(limit) || limit <= 0 || limit > 100) {
        throw httpError('limit must be an integer between 1 and 100', 400);
    }
    if (status && status !== 'all' && !VALID_STATUSES.has(status)) {
        throw httpError('Invalid purchase order status filter', 400);
    }
    if (!VALID_SORT_FIELDS.has(sortBy)) {
        throw httpError('Invalid sort field for purchase orders', 400);
    }
    if (sortOrder !== 'ASC' && sortOrder !== 'DESC') {
        throw httpError('sortOrder must be ASC or DESC', 400);
    }

    return purchaseRepo.getAllPurchaseOrders({
        page,
        limit,
        searchTerm: searchTerm.trim(),
        status,
        fromDate,
        toDate,
        sortBy,
        sortOrder,
    });
}

export function getPurchaseOrderById(id: number) {
    assertPositiveInteger(id, 'id');
    const po = purchaseRepo.getPurchaseOrderById(id);
    if (!po) throw httpError('Purchase Order not found', 404);
    return po;
}

export function createPurchaseOrder(data: CreatePOData) {
    assertPositiveInteger(data.supplier_id, 'supplier_id');
    assertPositiveInteger(data.created_by, 'created_by');

    if (!Array.isArray(data.items) || data.items.length === 0) {
        throw httpError('Purchase Order must have at least one item', 400);
    }

    const duplicateCheck = new Set<number>();
    data.items.forEach((item, index) => {
        assertPositiveInteger(item.variant_id, `items[${index}].variant_id`);
        assertPositiveInteger(item.quantity, `items[${index}].quantity`);
        if (!Number.isFinite(item.unit_cost) || item.unit_cost < 0) {
            throw httpError(`items[${index}].unit_cost must be a non-negative number`, 400);
        }
        if (duplicateCheck.has(item.variant_id)) {
            throw httpError(`Duplicate variant_id ${item.variant_id} is not allowed in a purchase order`, 400);
        }
        duplicateCheck.add(item.variant_id);
    });

    const poId = purchaseRepo.createPurchaseOrder(data);
    auditService.logCreate(data.created_by, 'purchase_orders', poId, {
        supplier_id: data.supplier_id,
        item_count: data.items.length,
        total_cost: data.items.reduce((sum, item) => sum + (item.quantity * item.unit_cost), 0),
    });
    return getPurchaseOrderById(poId);
}

export function updatePurchaseOrder(id: number, data: UpdatePOData) {
    assertPositiveInteger(id, 'id');
    assertPositiveInteger(data.supplier_id, 'supplier_id');
    assertPositiveInteger(data.updated_by, 'updated_by');

    const existingPo = getPurchaseOrderById(id);
    if (existingPo.status !== 'pending') {
        throw httpError('Only pending Purchase Orders can be updated', 409);
    }

    if (!Array.isArray(data.items) || data.items.length === 0) {
        throw httpError('Purchase Order must have at least one item', 400);
    }

    const duplicateCheck = new Set<number>();
    data.items.forEach((item, index) => {
        assertPositiveInteger(item.variant_id, `items[${index}].variant_id`);
        assertPositiveInteger(item.quantity, `items[${index}].quantity`);
        if (!Number.isFinite(item.unit_cost) || item.unit_cost < 0) {
            throw httpError(`items[${index}].unit_cost must be a non-negative number`, 400);
        }
        if (duplicateCheck.has(item.variant_id)) {
            throw httpError(`Duplicate variant_id ${item.variant_id} is not allowed in a purchase order`, 400);
        }
        duplicateCheck.add(item.variant_id);
    });

    purchaseRepo.updatePurchaseOrder(id, data);

    auditService.logUpdate(data.updated_by, 'purchase_orders', id,
        { supplier_id: existingPo.supplier_id, item_count: existingPo.items.length },
        { supplier_id: data.supplier_id, item_count: data.items.length }
    );

    return getPurchaseOrderById(id);
}

export function receivePurchaseOrder(id: number, userId: number) {
    assertPositiveInteger(id, 'id');
    assertPositiveInteger(userId, 'userId');

    const existingPo = getPurchaseOrderById(id);
    if (existingPo.status !== 'pending') {
        throw httpError('Only pending Purchase Orders can be received', 409);
    }

    purchaseRepo.receivePurchaseOrder(id, userId);
    const updatedPo = getPurchaseOrderById(id);

    auditService.logUpdate(
        userId,
        'purchase_orders',
        id,
        { status: existingPo.status, received_date: existingPo.received_date ?? null },
        { status: updatedPo.status, received_date: updatedPo.received_date ?? null },
    );

    return updatedPo;
}

export function cancelPurchaseOrder(id: number, userId: number) {
    assertPositiveInteger(id, 'id');
    assertPositiveInteger(userId, 'userId');

    const existingPo = getPurchaseOrderById(id);
    if (existingPo.status !== 'pending') {
        throw httpError('Only pending Purchase Orders can be cancelled', 409);
    }

    const result = purchaseRepo.cancelPurchaseOrder(id);
    if (result.changes === 0) {
        throw httpError('Cannot cancel Purchase Order. It may already be updated by another request.', 409);
    }

    const updatedPo = getPurchaseOrderById(id);
    auditService.logUpdate(
        userId,
        'purchase_orders',
        id,
        { status: existingPo.status },
        { status: updatedPo.status },
    );

    return updatedPo;
}

File: src/main/java/com/possum/application/purchase/PurchaseService.java
package com.possum.application.purchase;

import com.possum.domain.enums.FlowEventType;
import com.possum.domain.enums.PurchaseStatus;
import com.possum.domain.enums.TransactionType;
import com.possum.domain.exceptions.NotFoundException;
import com.possum.domain.exceptions.ValidationException;
import com.possum.domain.model.*;
import com.possum.infrastructure.serialization.JsonService;
import com.possum.persistence.db.ConnectionProvider;
import com.possum.persistence.db.TransactionManager;
import com.possum.persistence.repositories.interfaces.*;
import com.possum.shared.dto.PagedResult;
import com.possum.shared.dto.PurchaseOrderFilter;

import java.math.BigDecimal;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.time.LocalDateTime;
import java.util.*;

public class PurchaseService {
    private final PurchaseRepository purchaseRepository;
    private final SupplierRepository supplierRepository;
    private final VariantRepository variantRepository;
    private final InventoryRepository inventoryRepository;
    private final ProductFlowRepository productFlowRepository;
    private final AuditRepository auditRepository;
    private final TransactionManager transactionManager;
    private final ConnectionProvider connectionProvider;
    private final JsonService jsonService;

    public PurchaseService(PurchaseRepository purchaseRepository,
                           SupplierRepository supplierRepository,
                           VariantRepository variantRepository,
                           InventoryRepository inventoryRepository,
                           ProductFlowRepository productFlowRepository,
                           AuditRepository auditRepository,
                           TransactionManager transactionManager,
                           ConnectionProvider connectionProvider,
                           JsonService jsonService) {
        this.purchaseRepository = purchaseRepository;
        this.supplierRepository = supplierRepository;
        this.variantRepository = variantRepository;
        this.inventoryRepository = inventoryRepository;
        this.productFlowRepository = productFlowRepository;
        this.auditRepository = auditRepository;
        this.transactionManager = transactionManager;
        this.connectionProvider = connectionProvider;
        this.jsonService = jsonService;
    }

    public PagedResult<PurchaseOrder> getAllPurchaseOrders(PurchaseOrderFilter filter) {
        return purchaseRepository.getAllPurchaseOrders(filter);
    }

    public PurchaseOrderDetail getPurchaseOrderById(long id) {
        PurchaseOrder po = purchaseRepository.getPurchaseOrderById(id)
                .orElseThrow(() -> new NotFoundException("Purchase Order not found"));
        List<PurchaseOrderItem> items = purchaseRepository.getPurchaseOrderItems(id);
        return new PurchaseOrderDetail(po, items);
    }

    public PurchaseOrderDetail createPurchaseOrder(long supplierId, long createdBy, List<PurchaseOrderItemRequest> items) {
        com.possum.application.auth.ServiceSecurity.requirePermission(com.possum.application.auth.Permissions.PURCHASE_MANAGE);
        validateSupplier(supplierId);
        validateItems(items);

        return transactionManager.runInTransaction(() -> {
            List<PurchaseOrderItem> itemsToCreate = items.stream()
                    .map(req -> new PurchaseOrderItem(null, null, req.variantId(), null, null, null, req.quantity(), req.unitCost()))
                    .toList();

            long poId = purchaseRepository.createPurchaseOrder(supplierId, createdBy, itemsToCreate);

            BigDecimal totalCost = items.stream()
                    .map(item -> item.unitCost().multiply(BigDecimal.valueOf(item.quantity())))
                    .reduce(BigDecimal.ZERO, BigDecimal::add);

            Map<String, Object> auditData = Map.of(
                    "supplier_id", supplierId,
                    "item_count", items.size(),
                    "total_cost", totalCost
            );
            AuditLog auditLog = new AuditLog(null, createdBy, "CREATE", "purchase_orders", poId,
                    null, jsonService.toJson(auditData), null, null, LocalDateTime.now());
            auditRepository.insertAuditLog(auditLog);

            return getPurchaseOrderById(poId);
        });
    }

    public PurchaseOrderDetail updatePurchaseOrder(long id, long supplierId, long updatedBy, List<PurchaseOrderItemRequest> items) {
        com.possum.application.auth.ServiceSecurity.requirePermission(com.possum.application.auth.Permissions.PURCHASE_MANAGE);
        PurchaseOrderDetail existingPo = getPurchaseOrderById(id);
        if (!PurchaseStatus.PENDING.dbValue().equals(existingPo.purchaseOrder().status())) {
            throw new ValidationException("Only pending Purchase Orders can be updated");
        }

        validateSupplier(supplierId);
        validateItems(items);

        return transactionManager.runInTransaction(() -> {
            List<PurchaseOrderItem> itemsToUpdate = items.stream()
                    .map(req -> new PurchaseOrderItem(null, null, req.variantId(), null, null, null, req.quantity(), req.unitCost()))
                    .toList();

            purchaseRepository.updatePurchaseOrder(id, supplierId, itemsToUpdate);

            Map<String, Object> oldData = Map.of(
                    "supplier_id", existingPo.purchaseOrder().supplierId(),
                    "item_count", existingPo.items().size()
            );
            Map<String, Object> newData = Map.of(
                    "supplier_id", supplierId,
                    "item_count", items.size()
            );
            AuditLog auditLog = new AuditLog(null, updatedBy, "UPDATE", "purchase_orders", id,
                    jsonService.toJson(oldData), jsonService.toJson(newData), null, null, LocalDateTime.now());
            auditRepository.insertAuditLog(auditLog);

            return getPurchaseOrderById(id);
        });
    }

    public PurchaseOrderDetail receivePurchaseOrder(long id, long userId) {
        com.possum.application.auth.ServiceSecurity.requirePermission(com.possum.application.auth.Permissions.PURCHASE_MANAGE);
        PurchaseOrderDetail existingPo = getPurchaseOrderById(id);
        if (!PurchaseStatus.PENDING.dbValue().equals(existingPo.purchaseOrder().status())) {
            throw new ValidationException("Only pending Purchase Orders can be received");
        }

        if (existingPo.items().isEmpty()) {
            throw new ValidationException("Purchase Order has no items");
        }

        return transactionManager.runInTransaction(() -> {
            boolean updated = purchaseRepository.receivePurchaseOrder(id, userId);
            if (!updated) {
                throw new ValidationException("Purchase Order not found or already processed");
            }

            for (PurchaseOrderItem item : existingPo.items()) {
                if (item.quantity() <= 0 || item.unitCost().compareTo(BigDecimal.ZERO) < 0) {
                    throw new ValidationException("Purchase Order has invalid item data");
                }

                InventoryLot lot = new InventoryLot(null, item.variantId(), null, null, null,
                        item.quantity(), item.unitCost(), item.id(), LocalDateTime.now());
                long lotId = inventoryRepository.insertInventoryLot(lot);

                InventoryAdjustment adjustment = new InventoryAdjustment(null, item.variantId(), lotId,
                        item.quantity(), "confirm_receive", "purchase_order_item", item.id(),
                        userId, null, LocalDateTime.now());
                inventoryRepository.insertInventoryAdjustment(adjustment);

                ProductFlow flow = new ProductFlow(null, item.variantId(), FlowEventType.PURCHASE.getValue(),
                        item.quantity(), "purchase_order_item", item.id(), null, null, null, LocalDateTime.now());
                productFlowRepository.insertProductFlow(flow);
            }

            BigDecimal totalCost = existingPo.items().stream()
                    .map(item -> item.unitCost().multiply(BigDecimal.valueOf(item.quantity())))
                    .reduce(BigDecimal.ZERO, BigDecimal::add);

            if (totalCost.compareTo(BigDecimal.ZERO) > 0) {
                insertPurchaseTransaction(id, totalCost.negate());
            }

            Map<String, Object> oldData = Map.of(
                    "status", existingPo.purchaseOrder().status(),
                    "received_date", existingPo.purchaseOrder().receivedDate() != null ? existingPo.purchaseOrder().receivedDate() : ""
            );
            PurchaseOrderDetail updatedPo = getPurchaseOrderById(id);
            Map<String, Object> newData = Map.of(
                    "status", updatedPo.purchaseOrder().status(),
                    "received_date", updatedPo.purchaseOrder().receivedDate() != null ? updatedPo.purchaseOrder().receivedDate() : ""
            );
            AuditLog auditLog = new AuditLog(null, userId, "UPDATE", "purchase_orders", id,
                    jsonService.toJson(oldData), jsonService.toJson(newData), null, null, LocalDateTime.now());
            auditRepository.insertAuditLog(auditLog);

            return updatedPo;
        });
    }

    public PurchaseOrderDetail cancelPurchaseOrder(long id, long userId) {
        com.possum.application.auth.ServiceSecurity.requirePermission(com.possum.application.auth.Permissions.PURCHASE_MANAGE);
        PurchaseOrderDetail existingPo = getPurchaseOrderById(id);
        if (!PurchaseStatus.PENDING.dbValue().equals(existingPo.purchaseOrder().status())) {
            throw new ValidationException("Only pending Purchase Orders can be cancelled");
        }

        int rowsAffected = purchaseRepository.cancelPurchaseOrder(id);
        if (rowsAffected == 0) {
            throw new ValidationException("Cannot cancel Purchase Order. It may already be updated by another request.");
        }

        PurchaseOrderDetail updatedPo = getPurchaseOrderById(id);

        Map<String, Object> oldData = Map.of("status", existingPo.purchaseOrder().status());
        Map<String, Object> newData = Map.of("status", updatedPo.purchaseOrder().status());
        AuditLog auditLog = new AuditLog(null, userId, "UPDATE", "purchase_orders", id,
                jsonService.toJson(oldData), jsonService.toJson(newData), null, null, LocalDateTime.now());
        auditRepository.insertAuditLog(auditLog);

        return updatedPo;
    }

    private void validateSupplier(long supplierId) {
        supplierRepository.findSupplierById(supplierId)
                .orElseThrow(() -> new NotFoundException("Supplier " + supplierId + " not found"));
    }

    private void validateItems(List<PurchaseOrderItemRequest> items) {
        if (items == null || items.isEmpty()) {
            throw new ValidationException("Purchase Order must have at least one item");
        }

        Set<Long> variantIds = new HashSet<>();
        for (int i = 0; i < items.size(); i++) {
            PurchaseOrderItemRequest item = items.get(i);

            if (item.quantity() <= 0) {
                throw new ValidationException("items[" + i + "].quantity must be a positive integer");
            }

            if (item.unitCost().compareTo(BigDecimal.ZERO) < 0) {
                throw new ValidationException("items[" + i + "].unit_cost must be a non-negative number");
            }

            if (!variantIds.add(item.variantId())) {
                throw new ValidationException("Duplicate variant_id " + item.variantId() + " is not allowed in a purchase order");
            }

            if (variantRepository.findVariantByIdSync(item.variantId()).isEmpty()) {
                throw new NotFoundException("Variant " + item.variantId() + " not found");
            }
        }
    }

    private void insertPurchaseTransaction(long purchaseOrderId, BigDecimal amount) {
        Connection conn = connectionProvider.getConnection();
        try (PreparedStatement stmt = conn.prepareStatement(
                     "INSERT INTO transactions (purchase_order_id, amount, type, payment_method_id, status) VALUES (?, ?, ?, 1, 'completed')")) {
            stmt.setLong(1, purchaseOrderId);
            stmt.setBigDecimal(2, amount);
            stmt.setString(3, TransactionType.PURCHASE.dbValue());
            stmt.executeUpdate();
        } catch (Exception e) {
            throw new RuntimeException("Failed to insert purchase transaction", e);
        }
    }

    public record PurchaseOrderItemRequest(long variantId, int quantity, BigDecimal unitCost) {}
    public record PurchaseOrderDetail(PurchaseOrder purchaseOrder, List<PurchaseOrderItem> items) {}
}
